### DISCUSSION TITLE
Deleting annotations of a specific type

### DISCUSSION THREAD

**Discussion by qwertynik** [pymupdf/PyMuPDF#1672] [Looking for help] [answered]
Hello,

Used the below snippet to delete annotations, originally found [here](https://stackoverflow.com/questions/61947880/pymupdf-how-do-i-remove-annotations/62325900#62325900). 

```python
annot = page.firstAnnot
while annot:
    annot = page.delete_annot(annot)
```

When looking for ways to only delete specific annotations found that the `annot.type` property would help. However, noticed that the type is `Text` for both links and comments. Any ideas on how to delete annotations with greater precision? Also, could annotation deletion negatively impact other objects in the PDF?

Sample PDF [AnnotationTest1.pdf](https://github.com/pymupdf/PyMuPDF/files/8470197/AnnotationTest1.pdf)

Thanks!

**Comment by JorjMcKie** [comment, marked as answer]
Link, annotations and fields are handled as 3 different object types by MuPDF - although PDF uses the same technical basis for all of them.
So links and fields will not occur in iterator `page.annots()`. Similarly, only fields will occur in `page.widgets()`.
Your example has a link and a (text) annotation. The link will not be returned in neither `page.annots()` nor `page.delete_annot()`.
You could simply modify your above `while` statement: `while annot and annot.type[0] in (... applicable numeric annot types):`.

Or do this:
```python
alist = [a for a in page.annots(types=(...))]
for a in alist:
    page.delete_annot(a)
```

**Comment by qwertynik** [comment]
>Link, annotations and fields are handled as 3 different object types by MuPDF - although PDF uses the same technical basis for all of them.
So links and fields will not occur in iterator page.annots().

Thanks for this note @JorjMcKie, will certainly help in working more effectively with PDFs. 

Can you add clarity to this?
>Also, could annotation deletion negatively impact other objects in the PDF?

**Reply by JorjMcKie** [reply]
> could annotation deletion negatively impact other objects in the PDF?

No, the metaphor I like to use in this context:
Imagine a page to be some nice picture on the wall. Annotation are then like dust on that picture's glass cover: the picture content is not aware of the dust on it and not influenced by it. You can wipe away the dust (delete annotations), without changing the picture itself etc.

For a related reason, text highlighted by an annotation actually is a sloppy way of teeling the situation. You can add highlight annots without text (or anything at all for that matter) underneath. There is no connection - just coincident rectangles.

**Reply by JorjMcKie** [reply]
> could annotation deletion negatively impact other objects in the PDF?

No, the metaphor I like to use in this context:
Imagine a page to be some nice picture on the wall. Annotation are then like dust on that picture's glass cover: the picture content is not aware of the dust on it and not influenced by it. You can wipe away the dust (delete annotations), without changing the picture itself etc.

For a related reason, text highlighted by an annotation actually is a sloppy way of teeling the situation. You can add highlight annots without text (or anything at all for that matter) underneath. There is no connection - just coincident rectangles.

**Reply by qwertynik** [reply]
@JorjMcKie 

A highly useful and concise analogy.

Thanks! 🙏
